### PR TITLE
Add able to configure, add, remove Shells

### LINKED ISSUE #7523: Allow custom shell in Console tab
## Feature description

In the Settings dialog, I currently have a choice of bash, cmd, or powershell as the shell to run in the Console tab of the Browse dialog.

It would be nice if there were a way for us to specify the path to whatever shell we wanted, including command line arguments.

(I also think that PowerShell Core should be added as one of the presets.)

### PR DESCRIPTION
**Pull request by ivangrek**
Fixes #7523

## Proposed changes

- Allow custom shell in Console tab

### Before

image

### After

image

terminal

restart

image

## Test methodology 

- Manual
- Unit tests

## Test environment(s) 

- Git Extensions 3.4.3.9999
- Build d4b0f48bbf3e39a71f5d7ff5231be5678d4aa0f1
- Git 2.28.0.windows.1
- Microsoft Windows NT 10.0.19041.0
- .NET Framework 4.8.4220.0
- DPI 96dpi (no scaling)

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by RussKie** [comment]
This screenshot looks dated.
I may be dreaming, but I thought there was a discussion about breaking this
dialog and moving its elements to other pages. E.g. the style to the
Appearance page.
@pmiossec were you part to those discussions?

**Comment by pmiossec** [comment]
>@pmiossec were you part to those discussions?

That's this PR #8307
Perhaps we should merge it before working on this one...

**Comment by ivangrek** [comment]
This is the first step. The interface has not changed yet.

**Comment by ivangrek** [comment]
> This pull request has been automatically marked as stale because it has been marked as requiring author feedback but has not had any activity for 30 days. It will be closed if no further activity occurs.

Is the work still ongoing?

**Comment by pmiossec** [comment]
>Is the work still ongoing?

The PR was mostly finished. I just did the last ask modification. I hope it will be enough to see it merged...

**Comment by ivangrek** [comment]
Second commit:

image

---
`AppSettings.Saved += FillUserShells;`

I think we need something like this

AppSettings.[Value].Updated += FillUserShells;

Only if [New Value]  != [Old Value]

**Comment by ivangrek** [comment]
Third commit:

terminal

**Comment by ivangrek** [comment]
Need help.
I am trying to close my terminal session and it opens a console application.

image

**Comment by ivangrek** [comment]
> Add Shell refreshing after save settings

restart

File: GitUI/GitUI.csproj [hunk 1]
```diff
@@ -26,6 +26,7 @@
     <PackageReference Include="AppInsights.WindowsDesktop" Version="$(AppInsightsWindowsDesktopVersion)" />
     <PackageReference Include="Microsoft-WindowsAPICodePack-Core" Version="1.1.3.3" />
     <PackageReference Include="Microsoft-WindowsAPICodePack-Shell" Version="1.1.3.3" />
+    <PackageReference Include="Newtonsoft.Json" Version="10.0.3" />
```
Comment (RussKie) [inline GitUI/GitUI.csproj:29]: Shared references are configured via Solution.Versions.props
```suggestion
    <PackageReference Include="Newtonsoft.Json" Version="$(NewtonsoftJsonVersion)" />
```

File: GitUI/Configuring/Shell.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
+﻿namespace GitUI.Configuring
+{
+    public interface IShell
```
Comment (RussKie) [inline GitUI/Configuring/Shell.cs:3]: Why do we need this?
There's already `IShellDescriptor` and each shell has its own definition file.

**Review by RussKie — COMMENTED** [review]
I'm not quite sure we're moving in the right direction here. And I can't help it but the proposed API bring back horror memories of "diff/mergetools" implementation, sorry. The new infra appears to be duplicating the existing types, which I'm struggling to see the value in.

File: GitUI/Configuring/Shell.cs (hunk 2 above)
Comment (ivangrek) [reply GitUI/Configuring/Shell.cs:3]: Like in math. To simplify it is not always necessary to remove, sometimes it is necessary to add. 😄 
Refactoring later

**Comment by ivangrek** [comment]
Ready for review.

**Comment by ivangrek** [comment]
Any reviews?

**Comment by RussKie** [comment]
> Any reviews?

Yes, sorry, it's on my todo list...

**Comment by RussKie** [comment]
This is pretty cool, I couldn't get it working for the Scripts tab (🤦‍♂️ I know), but why is the list horizontal not a vertical?
image

**Comment by RussKie** [comment]
This is not correct - clicking to browse opens a wrong folder:
image

File: GitUI/CommandsDialogs/SettingsDialog/Pages/FormBrowseRepoSettingsPage.cs [hunk 3]
```diff
@@ -1,78 +1,502 @@
...
+
+            public override StandardValuesCollection GetStandardValues(ITypeDescriptorContext context)
+            {
+                var value = (string)context.PropertyDescriptor.GetValue(context.Instance);
+
+                return new StandardValuesCollection(ResourceImagesProvider.Images.Select(x => x.Key).ToArray());
+            }
+        }
+
+        private sealed class ResourceImagesTypeEditor : UITypeEditor
```
Comment (RussKie) [inline GitUI/CommandsDialogs/SettingsDialog/Pages/FormBrowseRepoSettingsPage.cs:228]: One class per file. Have a look at ScriptsSettingsPage.ScriptInfoProxy.cs.

File: GitUI/CommandsDialogs/SettingsDialog/Pages/FormBrowseRepoSettingsPage.cs [hunk 4]
```diff
@@ -1,78 +1,502 @@
...
+                    return listView.SelectedItems[0].Text;
+                }
+
+                return value;
+            }
         }
 
-        private void cboTerminal_Enter(object sender, EventArgs e)
+        [TypeConverter(typeof(PropertySorter))]
+        private sealed class ShellViewModel : IShell, INotifyPropertyChanged, ICheckable
```
Comment (RussKie) [inline GitUI/CommandsDialogs/SettingsDialog/Pages/FormBrowseRepoSettingsPage.cs:303]: We don't realy use View-ViewModel-Model naming convention. I suppose we can, but we'll need to be consistent and rename `ScriptInfoProxy`.

File: GitUI/CommandsDialogs/SettingsDialog/Pages/FormBrowseRepoSettingsPage.cs [hunk 5]
```diff
@@ -1,78 +1,502 @@
...
+                set
+                {
+                    if (_name == value)
+                    {
+                        return;
+                    }
+
+                    _name = value;
+
+                    PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Name)));
```
Comment (RussKie) [inline GitUI/CommandsDialogs/SettingsDialog/Pages/FormBrowseRepoSettingsPage.cs:367]: These can be made significantly simpler with the use of `CallerMemberName` attribute, have a look at https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.inotifypropertychanged?view=netframework-4.6.1#examples

File: GitUI/CommandsDialogs/SettingsDialog/Pages/FormBrowseRepoSettingsPage.cs [hunk 6]
```diff
@@ -1,78 +1,502 @@
...
+            [PropertyOrder(1)]
+            [Editor(typeof(ResourceImagesTypeEditor), typeof(UITypeEditor))]
+            [TypeConverter(typeof(ResourceImagesTypeConverter))]
+            public string Icon
+            {
+                get
+                {
+                    return _icon;
+                }
+
```
Comment (RussKie) [inline GitUI/CommandsDialogs/SettingsDialog/Pages/FormBrowseRepoSettingsPage.cs:381]: [nit] unnecessary blank line
[nit] convert to expressions, e.g. `get => _icon;`

File: GitUI/CommandsDialogs/SettingsDialog/Pages/FormBrowseRepoSettingsPage.cs [hunk 7]
```diff
@@ -1,78 +1,502 @@
...
         {
-            _cboTerminalPreviousIndex = cboTerminal.SelectedIndex;
+            private const string ShellCategory = "Shell";
+
+            public static ShellViewModel Create()
+            {
+                return new ShellViewModel();
+            }
+
+            public static ShellViewModel Create(IShell shell)
```
Comment (RussKie) [inline GitUI/CommandsDialogs/SettingsDialog/Pages/FormBrowseRepoSettingsPage.cs:312]: Use `implicit operator` instead

File: GitUI/CommandsDialogs/SettingsDialog/Pages/FormBrowseRepoSettingsPage.cs [hunk 8]
```diff
@@ -1,78 +1,502 @@
...
+                _arguments = ShellConstants.DefaultArguments;
+                _default = false;
+                _enabled = true;
+            }
+
+            [Browsable(false)]
+            public Guid Id { get; private set; }
+
+            [Category(ShellCategory)]
+            [PropertyOrder(0)]
```
Comment (RussKie) [inline GitUI/CommandsDialogs/SettingsDialog/Pages/FormBrowseRepoSettingsPage.cs:350]: I suggest we make the order consistent with the order of properties in the Script view, i.e.
image

**Review by RussKie — COMMENTED** [review]
Some high-level review.
Overall I think it is a good thing, though I'm not very fond of the new API, feels unnecessary verbose. I'd like us to ponder on it a bit more.

File: GitCommands/Settings/AppSettings.cs [hunk 9]
```diff
@@ -403,11 +405,14 @@ public static bool ShowResetAllChanges
...
-        public static ISetting<string> ConEmuFontSize => Setting.Create(DetailedSettingsPath, nameof(ConEmuFontSize), "12");
-        public static ISetting<bool> ShowGpgInformation => Setting.Create(DetailedSettingsPath, nameof(ShowGpgInformation), true);
+        public static readonly ISetting<bool> ShowConEmuTab = Setting.Create(DetailedSettingsPath, nameof(ShowConEmuTab), true);
+        public static readonly ISetting<string> ConEmuStyle = Setting.Create(DetailedSettingsPath, nameof(ConEmuStyle), "<Solarized Light>");
+        public static readonly ISetting<string> ConEmuFontSize = Setting.Create(DetailedSettingsPath, nameof(ConEmuFontSize), "12");
+        public static readonly ISetting<bool> ShowGpgInformation = Setting.Create(DetailedSettingsPath, nameof(ShowGpgInformation), true);
+
+        // Obsolete
+        public static readonly ISetting<string> ConEmuTerminal = Setting.Create(DetailedSettingsPath, nameof(ConEmuTerminal), "bash");
+        public static readonly ISetting<string> Shells = Setting.Create(DetailedSettingsPath, nameof(Shells), ShellsDefault);
```
[the remark this answers is not in the document]
Comment (ivangrek) [reply GitCommands/Settings/AppSettings.cs:415]: For migration, if we want to get the currently selected terminal.

File: GitUI/CommandsDialogs/SettingsDialog/Pages/FormBrowseRepoSettingsPage.cs (hunk 4 above)
Comment (ivangrek) [reply GitUI/CommandsDialogs/SettingsDialog/Pages/FormBrowseRepoSettingsPage.cs:303]: I think this is not really a proxy

File: GitUI/CommandsDialogs/SettingsDialog/Pages/FormBrowseRepoSettingsPage.cs (hunk 7 above)
Comment (ivangrek) [reply GitUI/CommandsDialogs/SettingsDialog/Pages/FormBrowseRepoSettingsPage.cs:312]: User-defined conversions are not allowed to convert from or to interface-types. In particular, this restriction ensures that no user-defined transformations occur when converting to an interface-type, and that a conversion to an interface-type succeeds only if the object being converted actually implements the specified interface-type.

File: GitUI/CommandsDialogs/SettingsDialog/Pages/FormBrowseRepoSettingsPage.ResourceImagesTypeConverter.cs [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+
+using System.ComponentModel;
+using System.Linq;
+using GitUI.Shells;
+
+namespace GitUI.CommandsDialogs.SettingsDialog.Pages
+{
+    public partial class FormBrowseRepoSettingsPage
+    {
+        private sealed class ResourceImagesTypeConverter : StringConverter
```
Comment (RussKie) [inline GitUI/CommandsDialogs/SettingsDialog/Pages/FormBrowseRepoSettingsPage.ResourceImagesTypeConverter.cs:11]: `ResourceImagesTypeConverter` and `ResourceImagesTypeEditor` need to be moved out of `FormBrowse*` namespace, they are general purpose infra, that we should be able to leverage for other things, like the Scripts.

**Review by RussKie — COMMENTED** [review]
A partial review of the plumbing changes, I'm yet to formally review the actual shell-related design and implementations.

**Pull request closed without merging**